File: packages/cli/src/commands/whoami.ts
import type { Command } from 'commander';
import { c, log, logError, logJson, getCurrentMode } from '../lib/output.js';
import { loadConfig } from '../lib/config.js';
import { createApiClient, ApiClientError } from '../lib/api-client.js';

export const registerWhoamiCommand = (program: Command): void => {
  program
    .command('whoami')
    .description('Display the currently logged-in user')
    .action(async () => {
      const config = loadConfig();

      if (!config.token) {
        if (getCurrentMode() === 'json') {
          logJson({ error: 'not_logged_in' });
          process.exitCode = 1;
          return;
        }
        logError(
          'Not logged in.',
          undefined,
          `Run ${c.cmd('spm login')} to authenticate with GitHub.`,
        );
        process.exitCode = 1;
        return;
      }

      const client = createApiClient(config);

      let user: Awaited<ReturnType<typeof client.whoami>>['user'];
      try {
        const res = await client.whoami();
        user = res.user;
      } catch (err) {
        if (err instanceof ApiClientError) {
          if (err.status === 401) {
            if (getCurrentMode() === 'json') {
              logJson({ error: 'token_expired' });
              process.exitCode = 1;
              return;
            }
            logError(
              'Session expired.',
              'Your authentication token is no longer valid.',
              `Run ${c.cmd('spm login')} to re-authenticate.`,
            );
            process.exitCode = 1;
            return;
          }
          logError('Could not reach registry', err.apiError.message);
          process.exitCode = 1;
          return;
        }
        throw err;
      }

      if (getCurrentMode() === 'json') {
        logJson({
          username: user.username,
          github: user.github_login,
          trust_tier: user.trust_tier,
          registered_at: user.created_at,
          skills_published: user.skills_published,
          registry: config.registry,
        });
        return;
      }

      console.log('');
      log(c.bold(user.username));
      log(`  GitHub:     github.com/${user.github_login}`);
      log(`  Trust tier: ${c.trust(user.trust_tier)}`);
      log(`  Registered: ${user.created_at}`);
      log(`  Published:  ${user.skills_published} skills`);
      log(`  Registry:   ${c.dim(config.registry)}`);
    });
};

File: cli-go/cmd/whoami.go
package cmd

import (
	"fmt"

	"github.com/skillpkg/spm/internal/api"
	"github.com/skillpkg/spm/internal/output"
	"github.com/spf13/cobra"
)

var whoamiCmd = &cobra.Command{
	Use:   "whoami",
	Short: "Display the currently logged-in user",
	Long:  "Show the username, GitHub login, trust tier, and other details for the authenticated user.",
	RunE:  runWhoami,
}

func init() {
	rootCmd.AddCommand(whoamiCmd)
}

func runWhoami(cmd *cobra.Command, _ []string) error {
	if Cfg.Token == "" {
		if Out.Mode == output.ModeJSON {
			_ = Out.LogJSON(map[string]string{"error": "not_logged_in"})
			return fmt.Errorf("not logged in")
		}
		Out.LogError("Not logged in.")
		Out.Log("Run %s to authenticate with GitHub.", output.Cyan("spm login"))
		return fmt.Errorf("not logged in")
	}

	client := api.NewClient(Cfg.RegistryURL(), Cfg.Token)
	user, err := client.Whoami()
	if err != nil {
		var apiErr *api.APIError
		if isAPIError(err, &apiErr) {
			if apiErr.IsUnauthorized() {
				if Out.Mode == output.ModeJSON {
					_ = Out.LogJSON(map[string]string{"error": "token_expired"})
					return fmt.Errorf("session expired")
				}
				Out.LogError("Session expired.")
				Out.Log("Your authentication token is no longer valid.")
				Out.Log("Run %s to re-authenticate.", output.Cyan("spm login"))
				return fmt.Errorf("session expired")
			}
			Out.LogError("Could not reach registry: %s", apiErr.Message)
			return fmt.Errorf("API error: %s", apiErr.Message)
		}
		return err
	}

	if Out.Mode == output.ModeJSON {
		return Out.LogJSON(map[string]any{
			"username":         user.Username,
			"github":           user.GithubLogin,
			"email":            user.Email,
			"trust_tier":       user.TrustTier,
			"registered_at":    user.CreatedAt,
			"skills_published": user.SkillsPublished,
			"registry":         Cfg.RegistryURL(),
		})
	}

	Out.Log("")
	Out.Log("%s", output.Bold(user.Username))
	Out.Log("  GitHub:     github.com/%s", user.GithubLogin)
	if user.Email != "" {
		Out.Log("  Email:      %s", user.Email)
	}
	Out.Log("  Trust tier: %s", user.TrustTier)
	Out.Log("  Registered: %s", user.CreatedAt)
	Out.Log("  Published:  %d skills", user.SkillsPublished)
	Out.Log("  Registry:   %s", output.Dim(Cfg.RegistryURL()))

	return nil
}
